### ISSUE TITLE
Add order by field into update query.

### ISSUE THREAD

**Issue by aleks-devel** [prisma/prisma#20301]
## Problem

When using increment or decrement, the order of change is important. Because if you start adding a number or subtracting from the wrong side, it can cause an error on unique indexes.

## Suggested solution

Add order by field for updateMany query

## Alternatives

Only raw queries

## Example:

```JavaScript
client.tableName.updateMany({
  where: {
    AND: [
      {
        orderIndex: { gte: 1 },
      },
      {
        orderIndex: {  lt: 5 },
      },
    ],
  },
  data: {
    orderIndex: {
      increment: 1,
    },
  },
});
```

Roughly translate to:

```SQL
UPDATE `table_name` SET `order_index` = (`table_name`.`order_index` + 1) WHERE (`table_name`.`order_index` >= 1 AND `table_name`.`order_index` < 5);
```

And this query will cause an error, because order_index is a unique index, and in the first updated row, a two will be set in a value that conflicts with the next line, which also has two.

The correct query should look like this, where the increment of the index starts from the other end, because of which there are no conflicts

```SQL
UPDATE `amberrie`.`social_networks` SET `order_index` = (`amberrie`.`social_networks`.`order_index` + 1) WHERE (`amberrie`.`social_networks`.`order_index` >= 1 AND `amberrie`.`social_networks`.`order_index` < 5) ORDER BY id DESC;
```

Or how it should look on prisma:

```JavaScript
client.tableName.updateMany({
  where: {
    AND: [
      {
        orderIndex: { gte: 1 },
      },
      {
        orderIndex: {
          lt: 5,
        },
      },
    ],
  },
  data: {
    orderIndex: {
      increment: 1,
    },
  },
  orderBy: {
    id: "desc"
  }
});
```

I understand that this situation occurs quite rarely, but still it is a necessary feature for me. Hope it gets added sometime

**Comment by nicksan222** [comment]
As a workaround you could first query in a defined order, then make a batch update

#### Ordered select
```ts
let dataToUpdate = await client.tableName.findMany({
  where: {
    AND: [
      {
        orderIndex: {
          gte: 1,
        },
      },
      {
        orderIndex: {
          lt: 5,
        },
      },
    ],
  },
  orderBy: {
    id: "desc"
  },
});
```

#### Batch update
```ts
await prisma.$transaction([
  dataToUpdate.map((row) => await client.tableName.update({
    where: {
      id: row.id,
    },
    data: {
      orderIndex: {
        increment: 1,
      },
    },
  });
  )
])

```

**Comment by daiho-nfq** [comment]
I'm not sure this is a bug or expected behaviour of `orderBy`

Ex:
Suppose there is a schema like below:
```
user one-to-many post

user (id, name)
post (id, user_id, status) // status enum: active and inactive
```

================================
CASE 1

user A has 5 posts
user B has 3 posts

my query
```
const result = this.prisma.user.findMany({
   orderBy: {
        posts: {
          _count: 'desc',
        },
      },
})
```

> Expected result will be user A -> user B => it works perfectly

================================
CASE 2

user A has 5 posts (1 active posts, 4 inactive posts)
user B has 3 posts (3 active posts, 0 inactive posts)

my query
```
const result = this.prisma.user.findMany({
  where: {
     posts: {
       some: {
         status: 'ACTIVE' // find users that have active posts
     }
    } 
  }
   select: {
     _count: {
       select: {
        posts: {
          status: 'ACTIVE' // count active post by users
       }
     }
    }
  }
   orderBy: {
        posts: {
          _count: 'desc',
        },
      },
})
```

> Expected result will be user B -> user A because user B has more active posts (3) than user A (1)

Howerver, the result is user A -> user B. As far as I inspect that `orderBy` is just order basing on all posts of user, no matter the `where` condition or the real `_count`

**Comment by aleks-devel** [comment]
@nicksan222 Yes, I'm already using this approach instead of raw queries. Thank you

**Comment by pascal-arnoux** [comment]
Hello,
I am also facing this prisma limitation where sql allows to update multiple rows in a given order to allow the transcient inconsistency triggerring a constraint failure (sql checks the contraints after every line update, not once all impacted lines are updated - reason why the ordered udpate is important and natively proposed by sql).
The only viable solution is to use a raw query today - a loop of queries is not acceptable in my case where I would need to convert a simple update ordered query into 1000s of queries.
Therefore, I am very interested in adding the "orderBy" capability to "updateMany" to avoid using a raw query.
Many thanks.

**Comment by basememara** [comment]
hitting this limitation as well, especially where we use `updateManyAndReturn` as a mechanism for handling concurrency. adding deterministic capability to the update many APIs would be a big unlock.
